backend/src/akvo/lumen/lib/aggregation/scatter.clj

```diff
@@ -1,102 +1,105 @@
-                                       " AS size, "
-                                       (sql-aggregation-subquery aggregation-method "%4$s" column-category-type)
-                                       " AS category, %10$s AS label FROM (SELECT * FROM %5$s WHERE %6$s ORDER BY random() LIMIT %8$s)z GROUP BY %10$s")
-        sql-text-without-aggregation "
-          SELECT * FROM (SELECT * FROM (SELECT %1$s AS x, %2$s AS y, %3$s AS size, %4$s AS category, %9$s AS label FROM %5$s WHERE %6$s)z ORDER BY random() LIMIT %8$s)zz ORDER BY zz.x"
-        sql-text (if have-aggregation sql-text-with-aggregation sql-text-without-aggregation)
-        sql-response (run-query tenant-conn table-name sql-text column-x-name column-y-name column-size-name  column-category-name filter-sql aggregation-method max-points column-label-name column-bucket-name)]
+        aggregation (partial sql-aggregation-subquery (:metricAggregation query))
+
+        subquery (format "(SELECT * FROM %1$s WHERE %2$s ORDER BY random() LIMIT %3$s)z"
+                         table-name filter-sql max-points)
+
+        sql-text-with-aggregation
+        (format "SELECT %1$s AS x, %2$s AS y, %3$s AS size, %4$s AS category, %5$s AS label 
+                 FROM %6$s
+                 GROUP BY %5$s"
+                (aggregation column-x)
+                (aggregation column-y)
+                (aggregation column-size)
+                (aggregation column-category)
+                (:columnName column-bucket)
+                subquery)
+        sql-text-without-aggregation (format "SELECT * FROM
+                                               (SELECT * FROM 
+                                                 (SELECT %1$s AS x, %2$s AS y, %3$s AS size, %4$s AS category, %5$s AS label 
+                                                  FROM %6$s 
+                                                  WHERE %7$s)z
+                                                ORDER BY random() 
+                                                LIMIT %8$s)zz
+                                              ORDER BY zz.x"
+                         (:columnName column-x)
+                         (:columnName column-y)
+                         (:columnName column-size)
+                         (:columnName column-category)
+                         (:columnName column-label)
+                         table-name
+                         filter-sql
+                         max-points)
+
+        sql-text (if column-bucket sql-text-with-aggregation sql-text-without-aggregation)
+        sql-response (run-query tenant-conn sql-text)]
     (lib/ok
-     {"series" (conj [{"key" column-x-title
-                        "label" column-x-title
-                        "data" (mapv (fn [[x-value y-value size-value category-value label]]
-                                        {"value" x-value})
+     {:series (conj [{:key (:title column-x)
```

Comment:
The creation of this map is duplicated per each column. Remove duplication

Reply:
is not really duplicated, in each column we take a different value ...

would you prefer this solution?


```clojure
{:series (conj [{:key      (:title column-x)
                      :label    (:title column-x)
                      :data     (mapv (fn [[x-value]] {:value x-value}) sql-response)
                      :metadata {:type (:type column-x)}}
                     {:key      (:title column-y)
                      :label    (:title column-y)
                      :data     (mapv (fn [[_ y-value]] {:value y-value}) sql-response)
                      :metadata {:type (:type column-y)}}]
                    (when (:title column-size) 
                      {:key      (:title column-size)
                       :label    (:title column-size)
                       :data     (mapv (fn [[_ _ size-value]] {:value size-value}) sql-response)
                       :metadata {:type (:type column-size)}})
                    (when (:title column-category)
                      {:key      (:title column-category)
                       :label    (:title column-category)
                       :data     (mapv (fn [[_ _ _ category-value]] {:value category-value}) sql-response)
                       :metadata {:type (:type column-category)}}))
      :common {:metadata {:type    (:type column-label)
                          :sampled (= (count sql-response) max-points)}
               :data     (mapv (fn [[_ _ _ _ label]] {:label label}) sql-response)}}

```
I know that we could _over_design a function that generates the 5 vectors in one iteration but not sure if we'll loose clarity here 🤔

Reply:
@dlebrero any thoughts on that ^^ ?

Reply:
Not complaining about iterating the same collection 5 times.

My initial thought was to just extract a fn so that the conj would look like:
(conj [(new-fn column-x 0) (new-fn column-y 1) (new-fn column-size 2) ...]

But maybe we can change the whole conj for a `(map-indexed new-fn [column-x column-y column-...]`

Reply:
what do you think about this approach?

```clojure
(let [sql-data [[:x0 :y0 :size0 :category0 :label0]
                [:x1 :y1 :size1 :category1 :label1]
                [:x2 :y2 :size2 :category2 :label2]]
      columns [{:title "x" :type  "text"}
               {:title "y" :type  "text"}
               {:title "size" :type  "text"}
               {:title "category" :type  "text"}
               {:title "label" :type  "text"}]]
  (->> sql-data
       (reduce
        (fn [c v]
          (-> c
              (update 0 conj (nth v 0))
              (update 1 conj (nth v 1))
              (update 2 conj (nth v 2))
              (update 3 conj (nth v 3))
              (update 4 conj (nth v 4))))
        [[] [] [] [] []])
       (map #(hash-map :key      (:title %),
                       :label    (:title %),
                       :data     %2,
                       :metadata {:type (:type %)})
            columns)))


;; ({:key "x",
;;   :label "x",
;;   :metadata {:type "text"},
;;   :data [:x0 :x1 :x2]}
;;  {:key "y",
;;   :label "y",
;;   :metadata {:type "text"},
;;   :data [:y0 :y1 :y2]}
;;  {:key "size",
;;   :label "size",
;;   :metadata {:type "text"},
;;   :data [:size0 :size1 :size2]}
;;  {:key "category",
;;   :label "category",
;;   :metadata {:type "text"},
;;   :data [:category0 :category1 :category2]}
;;  {:key "label",
;;   :label "label",
;;   :metadata {:type "text"},
;;   :data [:label0 :label1 :label2]})

```

Reply:
Pushed a commit with what I had in mind